### PR TITLE
stylo: ensure consistent order for custom properties computed values

### PR DESCRIPTION
**Pull request by ferjm**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

From https://bugzilla.mozilla.org/show_bug.cgi?id=1379577

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17973)

### CODE REVIEW

File: components/style/custom_properties.rs [hunk 1]
```diff
@@ -97,52 +98,128 @@ impl ToCss for ComputedValue {
...
+trait Iterable<K, V> {
+    fn index(&self) -> &Vec<K>;
+    fn values(&self) -> &HashMap<K, V>;
+}
+
+impl<K, V> Iterable<K, V> for OrderedMap<K, V>
+where
+    K: Eq + Hash + Clone,
+{
+    fn index(&self) -> &Vec<K> {
```
Comment (emilio) [inline components/style/custom_properties.rs]: Why is this trait needed? In Rust fields are private per-module, so you can use direct member access down below instead I think.

File: components/style/custom_properties.rs [hunk 2]
```diff
@@ -97,32 +97,35 @@ impl ToCss for ComputedValue {
...
 /// determined on the order that they are added to the name-value map.
+pub type CustomPropertiesMap = OrderedMap<ComputedValue>;
+
+/// A map that preserves order for the keys, and that is easily indexable.
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct CustomPropertiesMap {
+pub struct OrderedMap<T> {
     /// Custom property name index.
     index: Vec<Name>,
     /// Computed values indexed by custom property name.
```
Comment (heycam) [inline components/style/custom_properties.rs]: Please update the comments here not to talk about custom properties.

File: components/style/custom_properties.rs [hunk 3]
```diff
@@ -97,18 +98,24 @@ impl ToCss for ComputedValue {
...
-    values: HashMap<Name, T>,
+    values: HashMap<K, V>,
 }
 
-impl<T> OrderedMap<T> {
+impl<K, V> OrderedMap<K, V>
+where
+    K: Eq + Hash + Clone,
+{
     /// Creates a new custom properties map.
```
Comment (heycam) [inline components/style/custom_properties.rs]: And all these comments too.

File: components/style/custom_properties.rs [hunk 4]
```diff
@@ -168,6 +172,38 @@ where
     }
 }
 
+/// An iterator for OrderedMap
```
Comment (heycam) [inline components/style/custom_properties.rs]: Can you make the order that we iterate over the items clear in this comment?

File: components/style/custom_properties.rs [hunk 5]
```diff
@@ -168,6 +172,38 @@ where
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ref index = self.inner.index;
+        if self.pos >= index.len() {
+            return None;
+        }
+
+        let ref name = index[index.len() - self.pos - 1];
+        self.pos += 1;
+        match self.inner.values.get(name) {
```
Comment (heycam) [inline components/style/custom_properties.rs]: It should be an error if we get None from this, right?  Can we just unwrap() instead of returning None?

**Comment by heycam** [comment]
@bors-servo delegate+

r=me with those issues addressed.  (Some of the comment-related ones seem to be fixed in the later commits; don't bother moving them to the earlier commits now.)

**Pull request merged**